<a href="https://colab.research.google.com/github/Jojomaque/Portfolio/blob/main/Jojo_Preparing_data_for_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Preparing Data for Analysis**



*   Exercise 1 - Checking for null vlues
*   Exercise 2 - Removing null values
*   Exercise 3 - Removing duplicate records from a data set
*   Exercise 4 - Normalising values where necessary
*   Exercise 5 - Correcting the format of dates where necessary
*   Exercise 6 - Merging & Joining data + Joining data from 2 different sources
*   Exercise 7 - Resetting data values











### Exercise 1 - check for null values 
---

1. Read data from the file housing_in_london_yearly_variables.csv from this link: https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv 
2. check if any NA values exist in the dataframe and print the result 
3. use df.info() to see which columns have null entries (*Hint: if the non-null count is less than total entries, column contains missing/NA entries*)  

**Test output**:
True
.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071) 



In [ ]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv')
housing_london_original = df
housing_london_original.isnull().values.any()

True

In [ ]:
#.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071)
housing_london_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1071 entries, 0 to 1070
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               1071 non-null   object 
 1   area               1071 non-null   object 
 2   date               1071 non-null   object 
 3   median_salary      1049 non-null   float64
 4   life_satisfaction  352 non-null    float64
 5   mean_salary        1071 non-null   object 
 6   recycling_pct      860 non-null    object 
 7   population_size    1018 non-null   float64
 8   number_of_jobs     931 non-null    float64
 9   area_size          666 non-null    float64
 10  no_of_houses       666 non-null    float64
 11  borough_flag       1071 non-null   int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 100.5+ KB


### Exercise 2 - remove null values 
---

1. Remove rows with NA values for `life_satisfaction` (use [ ] even if only one column in list)
2. Remove all NA values across whole dataframe 

**Test output**:  
1.  Row count reduced to 352 rows
2.  Row count reduced to 267 rows

In [ ]:
#Remove rows with NA values for life_satisfaction
housing_london_original.dropna(subset = ['life_satisfaction'])

code                  area  ... no_of_houses  borough_flag
613   E09000002  barking and dagenham  ...      71079.0             1
614   E09000003                barnet  ...     139346.0             1
615   E09000004                bexley  ...      95037.0             1
616   E09000005                 brent  ...     112083.0             1
617   E09000006               bromley  ...     135036.0             1
...         ...                   ...  ...          ...           ...
1010  E12000009            south west  ...          NaN             0
1013  E92000001               england  ...   24172166.0             0
1017  N92000002      northern ireland  ...          NaN             0
1018  S92000003              scotland  ...          NaN             0
1019  W92000004                 wales  ...          NaN             0

[352 rows x 12 columns]

In [ ]:
#Remove rows with NA values for whola dataframe
housing_london_original.dropna()

code                  area  ... no_of_houses  borough_flag
613   E09000002  barking and dagenham  ...      71079.0             1
614   E09000003                barnet  ...     139346.0             1
615   E09000004                bexley  ...      95037.0             1
616   E09000005                 brent  ...     112083.0             1
617   E09000006               bromley  ...     135036.0             1
...         ...                   ...  ...          ...           ...
999   E09000031        waltham forest  ...     103029.0             1
1000  E09000032            wandsworth  ...     146162.0             1
1001  E09000033           westminster  ...     124509.0             1
1008  E12000007                london  ...    3556161.0             0
1013  E92000001               england  ...   24172166.0             0

[267 rows x 12 columns]

## Dropping duplicates
---

* To remove duplicate rows based on duplication of values in all columns  
  `df.drop_duplicates()`  

* To remove rows that have duplicate entries in a specified column  
  `df.drop_duplicates(subset = ['Make'])`  

* To remove rows that have duplicate entries in multiple columns  
  `df.drop_duplicates(subset = ['Make', 'Model'])` 

* Remove duplicate rows keeping the last instance rather than the first (default):  
  `df.drop_duplicates(keep='last')`  

### Exercise 3 - Removing duplicate entries 
---

remove duplicate `area` entries keeping first instance  

**Test output**:  
 Dataframe now contains 50 rows all with date 1999-12-*01* 

In [ ]:
#remove duplicate area entries keeping first instance
housing_london_original.drop_duplicates(subset = ['area'])


code                      area  ... no_of_houses  borough_flag
0   E09000001            city of london  ...          NaN             1
1   E09000002      barking and dagenham  ...          NaN             1
2   E09000003                    barnet  ...          NaN             1
3   E09000004                    bexley  ...          NaN             1
4   E09000005                     brent  ...          NaN             1
5   E09000006                   bromley  ...          NaN             1
6   E09000007                    camden  ...          NaN             1
7   E09000008                   croydon  ...          NaN             1
8   E09000009                    ealing  ...          NaN             1
9   E09000010                   enfield  ...          NaN             1
10  E09000011                 greenwich  ...          NaN             1
11  E09000012                   hackney  ...          NaN             1
12  E09000013    hammersmith and fulham  ...          NaN             1
13  E09000014                  haringey  ...          NaN             1
14  E09000015                    harrow  ...          NaN             1
15  E09000016                  havering  ...          NaN             1
16  E09000017                hillingdon  ...          NaN             1
17  E09000018                  hounslow  ...          NaN             1
18  E09000019                 islington  ...          NaN             1
19  E09000020    kensington and chelsea  ...          NaN             1
20  E09000021      kingston upon thames  ...          NaN             1
21  E09000022                   lambeth  ...          NaN             1
22  E09000023                  lewisham  ...          NaN             1
23  E09000024                    merton  ...          NaN             1
24  E09000025                    newham  ...          NaN             1
25  E09000026                 redbridge  ...          NaN             1
26  E09000027      richmond upon thames  ...          NaN             1
27  E09000028                 southwark  ...          NaN             1
28  E09000029                    sutton  ...          NaN             1
29  E09000030             tower hamlets  ...          NaN             1
30  E09000031            waltham forest  ...          NaN             1
31  E09000032                wandsworth  ...          NaN             1
32  E09000033               westminster  ...          NaN             1
33  E12000001                north east  ...          NaN             0
34  E12000002                north west  ...          NaN             0
35  E12000003  yorkshire and the humber  ...          NaN             0
36  E12000004             east midlands  ...          NaN             0
37  E12000005             west midlands  ...          NaN             0
38  E12000006                      east  ...          NaN             0
39  E12000007                    london  ...          NaN             0
40  E12000008                south east  ...          NaN             0
41  E12000009                south west  ...          NaN             0
42  E13000001              inner london  ...          NaN             0
43  E13000002              outer london  ...          NaN             0
44  E92000001                   england  ...          NaN             0
45  K02000001            united kingdom  ...          NaN             0
46  K03000001             great britain  ...          NaN             0
47  K04000001         england and wales  ...          NaN             0
48  N92000002          northern ireland  ...          NaN             0
49  S92000003                  scotland  ...          NaN             0
50  W92000004                     wales  ...          NaN             0

[51 rows x 12 columns]

 Exercise 4 - prepare dataset for normalisation 
---
1. Read Covid vaccination data from the `by_country` sheet in the Excel file at this link : https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/`Covid`%20Vaccination%20Data.xlsx?raw=true
2. Use .info() to find out which columns have missing values
3. Remove all rows with missing data in the total_vaccination column
4. Find the median vaccinations per hundred 
5. display the mean people vaccinated per hundred for each country in descending order
6. find the range of total_vaccinations across the dataframe 


**Test output**:  
1. dataframe is saved in a variable
2. 
```
RangeIndex: 14994 entries, 0 to 14993
Data columns (total 15 columns):
    Column                               Non-Null Count  Dtype         
                                
 0   country                              14994 non-null  object        
 1   iso_code                             14994 non-null  object        
 2   date                                 14994 non-null  datetime64[ns]
 3   total_vaccinations                   9011 non-null   float64       
 4   people_vaccinated                    8370 non-null   float64       
 5   people_fully_vaccinated              6158 non-null   float64       
 6   daily_vaccinations_raw               7575 non-null   float64       
 7   daily_vaccinations                   14796 non-null  float64       
 8   total_vaccinations_per_hundred       9011 non-null   float64       
 9   people_vaccinated_per_hundred        8370 non-null   float64       
 10  people_fully_vaccinated_per_hundred  6158 non-null   float64       
 11  daily_vaccinations_per_million       14796 non-null  float64       
 12  vaccines                             14994 non-null  object        
 13  source_name                          14994 non-null  object        
 14  source_website                       14994 non-null  object        
dtypes: datetime64[ns](1), float64(9), object(5)
memory usage: 1.7+ MB
```
3. 9011 rows × 15 columns
4. 7.78
5. 
```
country
Gibraltar       72.172462
Maldives        51.276087
Israel          48.931008
Seychelles      48.233333
Aruba           42.155000
                  ...    
Guinea           0.677000
Sierra Leone     0.530000
Namibia          0.317143
South Africa     0.261667
Albania          0.080000
Name: people_vaccinated_per_hundred, Length: 104, dtype: float64
```
6. 275338000.0


Exercise 4.1 - Read Covid vaccination data from the by_country sheet in the Excel file at this link 


In [ ]:
#importing libraries
import pandas as pd
#reading the data set from an excel file using the Pandas Library
df = pd.read_excel('https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true')
#Renaming the variable
dfvac = df
dfvac

country  ...                                     source_website
0      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
1      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
2      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
3      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
4      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
...            ...  ...                                                ...
14989     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14990     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14991     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14992     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14993     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...

[14994 rows x 15 columns]

Exercise 4.2 - Use .info() to find out which columns have missing values


In [ ]:
#Use .info() to find out which columns have missing values
dfvac.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14994 entries, 0 to 14993
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   country                              14994 non-null  object        
 1   iso_code                             14994 non-null  object        
 2   date                                 14994 non-null  datetime64[ns]
 3   total_vaccinations                   9011 non-null   float64       
 4   people_vaccinated                    8370 non-null   float64       
 5   people_fully_vaccinated              6158 non-null   float64       
 6   daily_vaccinations_raw               7575 non-null   float64       
 7   daily_vaccinations                   14796 non-null  float64       
 8   total_vaccinations_per_hundred       9011 non-null   float64       
 9   people_vaccinated_per_hundred        8370 non-null   float64       
 10  people_ful

Exercise 4.3 - Remove all rows with missing data in the total_vaccination column


In [5]:
df = pd.read_excel('https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true')
#Remove all rows with missing data in the total_vaccination column
dfvacc = df.dropna(subset = ['total_vaccinations'])
display(dfvacc.head())
display(dfvacc.info())

country  ...                                     source_website
0   Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
6   Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
22  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
44  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
59  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...

[5 rows x 15 columns]

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9011 entries, 0 to 14993
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   country                              9011 non-null   object        
 1   iso_code                             9011 non-null   object        
 2   date                                 9011 non-null   datetime64[ns]
 3   total_vaccinations                   9011 non-null   float64       
 4   people_vaccinated                    8294 non-null   float64       
 5   people_fully_vaccinated              6157 non-null   float64       
 6   daily_vaccinations_raw               7575 non-null   float64       
 7   daily_vaccinations                   8816 non-null   float64       
 8   total_vaccinations_per_hundred       9011 non-null   float64       
 9   people_vaccinated_per_hundred        8294 non-null   float64       
 10  people_full

None

Exercise 4.4 - Find the median vaccinations per hundred


In [3]:
import pandas as pd
dfvacc = pd.read_excel('https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true')
# Python code to show the working of median() function.
dfvacc['total_vaccinations_per_hundred'].median()





6.3



Exercise 4.5 - Display the mean people vaccinated per hundred for each country in descending order


In [6]:

import pandas as pd
dfvacc = pd.read_excel('https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true')
dfvacc.dropna(subset = ['people_vaccinated_per_hundred'])

dfvacc.groupby(['country'])['people_vaccinated_per_hundred',].mean().sort_values(['people_vaccinated_per_hundred'],ascending = False)

people_vaccinated_per_hundred
country                                                        
Gibraltar                                             64.975699
Bhutan                                                55.961892
Falkland Islands                                      51.063333
Saint Helena                                          44.880000
Seychelles                                            44.005686
...                                                         ...
China                                                       NaN
Ethiopia                                                    NaN
Saint Vincent and the Grenadines                            NaN
Samoa                                                       NaN
Saudi Arabia                                                NaN

[195 rows x 1 columns]

Exercise 4.6 - Find the range of total_vaccinations across the dataframe

In [7]:
#6 find the range of total_vaccinations across the dataframe

import pandas as pd
dfvacc = pd.read_excel('https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true')
dfvacc.dropna(subset=["total_vaccinations"])
range=dfvacc['total_vaccinations'].max()-dfvacc['total_vaccinations'].min()
range

275338000.0

### Challenge 2 - normalise daily vaccinations 
---

1. Find the median daily vaccinations per million 
2. Write a function to normalise daily vaccinations per million, where values greater than or equal to median = 1 and values less than median = 0 

**Test output**: 

1. 1915.5
2. 
```
0        0
6        0
22       0
44       0
59       0
        ..
14989    0
14990    0
14991    0
14992    0
14993    0
Name: daily_vaccinations_per_million, Length: 9011, dtype: int64
```

In [ ]:

display(dfvac2["daily_vaccinations_per_million"].median())


1915.5

In [ ]:
#Write a function to normalise daily vaccinations per million, where values greater than or equal to median = 1 and values less than median = 0
df = pd.read_excel('https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true')
df
daily_vac1 = df

daily_vac1 = daily_vac1.dropna(subset=["daily_vaccinations_per_million"])

def normalise_daily_vac(daily_vac1):
  if daily_vac1['daily_vaccinations_per_million'] >= 1475.0:
    return 1
  else:
    return 0

daily_vac1['daily_vaccinations_per_million'] = daily_vac1.apply(normalise_daily_vac, axis=1)
daily_vac1

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


country  ...                                     source_website
1      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
2      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
3      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
4      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
5      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
...            ...  ...                                                ...
14989     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14990     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14991     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14992     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14993     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...

[14796 rows x 15 columns]

In [ ]:
daily_vac1['daily_vaccinations_per_million'].max()

1

### Challenge 3 - Normalising total vaccinations   
---
The United Kingdom has been praised for its fast vaccine rollout. 
1. Find the minimum total vaccinations for the United Kingdom 
2. Save this value in a variable rounded down to an integer
3. Write a function to normalise total_vaccinations column so that all values less than the UK's min are 0 and all values greater than or equal to the UK's min are coded as 1 
4. Display the countries for which total vaccinated is at the same rate or more than the UK

**Test output**:

1. 1402432.0
2. 1402432
3. `df['total_vaccinations']` should output:
```
0        0
6        0
22       0
44       0
59       0
        ..
14989    0
14990    0
14991    0
14992    0
14993    0
Name: total_vaccinations, Length: 9011, dtype: int64
```
4. 
```
array(['Argentina', 'Australia', 'Austria', 'Azerbaijan', 'Bangladesh',
       'Belgium', 'Brazil', 'Cambodia', 'Canada', 'Chile', 'China',
       'Colombia', 'Czechia', 'Denmark', 'Dominican Republic', 'England',
       'Finland', 'France', 'Germany', 'Greece', 'Hong Kong', 'Hungary',
       'India', 'Indonesia', 'Ireland', 'Israel', 'Italy', 'Japan',
       'Kazakhstan', 'Malaysia', 'Mexico', 'Morocco', 'Nepal',
       'Netherlands', 'Norway', 'Pakistan', 'Peru', 'Philippines',
       'Poland', 'Portugal', 'Qatar', 'Romania', 'Russia', 'Saudi Arabia',
       'Scotland', 'Serbia', 'Singapore', 'Slovakia', 'South Korea',
       'Spain', 'Sweden', 'Switzerland', 'Thailand', 'Turkey',
       'United Arab Emirates', 'United Kingdom', 'United States',
       'Uruguay', 'Wales'], dtype=object)
```




In [ ]:
#Find the minimum total vaccinations for the United Kingdom
df = pd.read_excel('https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true')
#Filtering Rows of Pandas Dataframe 
df_uk = df[df.country == 'United Kingdom']
min_total_vac = df_uk['total_vaccinations'
].min()
min_total_vac

1402432.0

In [ ]:
min_total_vac_round = int(min_total_vac)
min_total_vac_round


1402432

In [ ]:
df = pd.read_excel('https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true')

df = df.dropna(subset=["total_vaccinations"])
def normalise_total_vac(df):
  if df['total_vaccinations'] < 1402432:
    return 0
  else:
    return 1

df['total_vaccinations'] = df.apply(normalise_total_vac, axis=1)
df['total_vaccinations']


0        0
6        0
22       0
44       0
59       0
        ..
14989    0
14990    0
14991    0
14992    0
14993    0
Name: total_vaccinations, Length: 9011, dtype: int64

In [ ]:
#Display the countries for which total vaccinated is at the same rate or more than the UK
df = pd.read_excel('https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true')

df = df.dropna(subset=["total_vaccinations_per_hundred"])

# df_rate
# result = df_country.loc[df_country['total_vaccinations_per_hundred'] >= 2748464937 , ['country']]
# result

country  ...                                     source_website
0      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
6      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
22     Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
44     Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
59     Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
...            ...  ...                                                ...
14989     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14990     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14991     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14992     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14993     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...

[8485 rows x 15 columns]

### Challenge 4 - create new series of total vaccinations for each manufacturer
---

To create a new column in your dataframe:

`df['new_column'] = ...`

For example:

* to duplicate an existing column
  * `df['new_column'] = df['old_column']`
* to add two columns together 
  * `df['new_column'] = df['column1'] + df['column2']`
* to make a percentages column 
  * `df['new_column'] = (df['column1']/df['column1].sum()) * 100`

  
1. read data from 'by_manufacturer' sheet from Covid data 
2. find the sum of total vaccinations for each manufacturer
3. create a new column that has the total vaccinations as a percentage of the overall sum of total vaccinations 
4. find the median percentage 
5. create a new column called 'normalised_percentages' which duplicates the percentages column
6. normalise the normalised_percentages column so that any values greater than or equal to the median percentage = 1 and any lesser than = 0 


**Test output**:

1.
2. 
```
vaccine
Johnson&Johnson        264839828
Moderna               5548036383
Oxford/AstraZeneca     539433203
Pfizer/BioNTech       8690461304
Sinovac                604660293
Name: total_vaccinations, dtype: int64
```
3. 
```
	location	date	vaccine	total_vaccinations	percentages
0	Chile	2020-12-24	Pfizer/BioNTech	420	0.000003
1	Chile	2020-12-25	Pfizer/BioNTech	5198	0.000033
2	Chile	2020-12-26	Pfizer/BioNTech	8338	0.000053
3	Chile	2020-12-27	Pfizer/BioNTech	8649	0.000055
4	Chile	2020-12-28	Pfizer/BioNTech	8649	0.000055
...	...	...	...	...	...
3291	United States	2021-05-01	Moderna	105947940	0.677095
3292	United States	2021-05-01	Pfizer/BioNTech	129013657	0.824504
3293	United States	2021-05-02	Johnson&Johnson	8374395	0.053519
3294	United States	2021-05-02	Moderna	106780082	0.682413
3295	United States	2021-05-02	Pfizer/BioNTech	130252779	0.832423
3296 rows × 5 columns
```
4. 0.0011110194374896931
5. 
6. 
```
	location	date	vaccine	total_vaccinations	percentages	normalise	normalised
0	Chile	2020-12-24	Pfizer/BioNTech	420	0.000003	0.000003	0
1	Chile	2020-12-25	Pfizer/BioNTech	5198	0.000033	0.000033	0
2	Chile	2020-12-26	Pfizer/BioNTech	8338	0.000053	0.000053	0
3	Chile	2020-12-27	Pfizer/BioNTech	8649	0.000055	0.000055	0
4	Chile	2020-12-28	Pfizer/BioNTech	8649	0.000055	0.000055	0
...	...	...	...	...	...	...	...
3291	United States	2021-05-01	Moderna	105947940	0.677095	0.677095	1
3292	United States	2021-05-01	Pfizer/BioNTech	129013657	0.824504	0.824504	1
3293	United States	2021-05-02	Johnson&Johnson	8374395	0.053519	0.053519	1
3294	United States	2021-05-02	Moderna	106780082	0.682413	0.682413	1
3295	United States	2021-05-02	Pfizer/BioNTech	130252779	0.832423	0.832423	1
3296 rows × 7 columns
```



In [ ]:
import pandas as pd
#read data from 'by_manufacturer' sheet from Covid data
df = pd.read_excel('https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true', sheet_name = 'by_manufacturer')
df

location       date          vaccine  total_vaccinations
0             Chile 2020-12-24  Pfizer/BioNTech                 420
1             Chile 2020-12-25  Pfizer/BioNTech                5198
2             Chile 2020-12-26  Pfizer/BioNTech                8338
3             Chile 2020-12-27  Pfizer/BioNTech                8649
4             Chile 2020-12-28  Pfizer/BioNTech                8649
...             ...        ...              ...                 ...
3291  United States 2021-05-01          Moderna           105947940
3292  United States 2021-05-01  Pfizer/BioNTech           129013657
3293  United States 2021-05-02  Johnson&Johnson             8374395
3294  United States 2021-05-02          Moderna           106780082
3295  United States 2021-05-02  Pfizer/BioNTech           130252779

[3296 rows x 4 columns]

In [ ]:
df_groupby_vaccine = df.groupby(['vaccine'])[['vaccine','total_vaccinations']].sum().reset_index()
df_groupby_vaccine

vaccine  total_vaccinations
0     Johnson&Johnson           264839828
1             Moderna          5548036383
2  Oxford/AstraZeneca           539433203
3     Pfizer/BioNTech          8690461304
4             Sinovac           604660293

In [ ]:
df = pd.read_excel('https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true', sheet_name = 'by_manufacturer')
df
#create a new column that has the total vaccinations as a percentage of the overall sum of total vaccinations
df['percent'] = (df['total_vaccinations']/df['total_vaccinations'].sum()) * 100
df


location       date          vaccine  total_vaccinations   percent
0             Chile 2020-12-24  Pfizer/BioNTech                 420  0.000003
1             Chile 2020-12-25  Pfizer/BioNTech                5198  0.000033
2             Chile 2020-12-26  Pfizer/BioNTech                8338  0.000053
3             Chile 2020-12-27  Pfizer/BioNTech                8649  0.000055
4             Chile 2020-12-28  Pfizer/BioNTech                8649  0.000055
...             ...        ...              ...                 ...       ...
3291  United States 2021-05-01          Moderna           105947940  0.677095
3292  United States 2021-05-01  Pfizer/BioNTech           129013657  0.824504
3293  United States 2021-05-02  Johnson&Johnson             8374395  0.053519
3294  United States 2021-05-02          Moderna           106780082  0.682413
3295  United States 2021-05-02  Pfizer/BioNTech           130252779  0.832423

[3296 rows x 5 columns]

In [ ]:
#find the median percentage
df['percent'].median()

0.0011110194374896931

In [ ]:
#create a new column called 'normalised_percentages' which duplicates the percentages column
df['normalised_percentages'] = df['percent']
df


location       date  ...   percent  normalised_percentages
0             Chile 2020-12-24  ...  0.000003                0.000003
1             Chile 2020-12-25  ...  0.000033                0.000033
2             Chile 2020-12-26  ...  0.000053                0.000053
3             Chile 2020-12-27  ...  0.000055                0.000055
4             Chile 2020-12-28  ...  0.000055                0.000055
...             ...        ...  ...       ...                     ...
3291  United States 2021-05-01  ...  0.677095                0.677095
3292  United States 2021-05-01  ...  0.824504                0.824504
3293  United States 2021-05-02  ...  0.053519                0.053519
3294  United States 2021-05-02  ...  0.682413                0.682413
3295  United States 2021-05-02  ...  0.832423                0.832423

[3296 rows x 6 columns]

In [ ]:
#normalise the normalised_percentages column so that any values greater than or equal to the median percentage = 1 and any lesser than = 0

def normalise_percentage(df):

  if df['percent'] >= 0.0011:
   return 1
  else:
   return 0
df['normalised_percentages'] = df.apply(normalise_percentage, axis=1)
df




location       date  ...   percent  normalised_percentages
0             Chile 2020-12-24  ...  0.000003                       0
1             Chile 2020-12-25  ...  0.000033                       0
2             Chile 2020-12-26  ...  0.000053                       0
3             Chile 2020-12-27  ...  0.000055                       0
4             Chile 2020-12-28  ...  0.000055                       0
...             ...        ...  ...       ...                     ...
3291  United States 2021-05-01  ...  0.677095                       1
3292  United States 2021-05-01  ...  0.824504                       1
3293  United States 2021-05-02  ...  0.053519                       1
3294  United States 2021-05-02  ...  0.682413                       1
3295  United States 2021-05-02  ...  0.832423                       1

[3296 rows x 6 columns]

Exercise 5 - Correcting the format of dates where necessary

---
### Exercise 5.1 - create a valid date

Write a function that will take 3 parameters, day, month, year (3 numbers) and will:  
*   check that the month is between 1 and 12  
*   check that the day is valid for the month (e.g. if the month is 2, then the day must be between 1 and 28 (don't worry about leap years for the moment)  
*   if the date is not valid, print an error message and return None (which the caller will print)  
*   if the date is valid, store it in a new variable, return the variable and the caller will print it.

In [ ]:
from datetime import date

def get_valid_date(day,month,year):
  #check that the month is between 1 and 12
  if month > 0 and month < 13:

  #check that the day is valid for the month (e.g. if the month is 2, then the day must be between 1 and 28 (don't worry about leap years for the moment)
    if day > 0:
      if month in [1, 3, 5 , 7, 8, 10, 12] and day < 32:
        return date(year, month, day)
      elif month in [4, 6, 9, 11] and day < 31:
        return date(year, month, day)
      elif month == 2 and day < 29:
        return date(year, month, day)
  
  return None

print(get_valid_date(31, 12, 2021))
print(get_valid_date(1, 13, 2021))
print(get_valid_date(31, 1, 2021))
print(get_valid_date(31, 2, 2021))

2021-12-31
None
2021-01-31
None


---
### Exercise 5.2 - is it a leap year?

Write a function that will take a date as a parameter (you will need to create the date variable before calling the function) and will determine if it is a leap year, returning True or False.

*Hint:  a leap year is a year that is divisible by 4 but the centuries (1900, 2000, etc are only leap years if they are divisible by 400)*

In [ ]:
from datetime import datetime
def is_leap_year(date_to_check):
  # add your code below here
  # find the year (int) for the date_to_check
  # return the condition check; this will be a boolean (True or False)
  year = date_to_check.year
  return year % 4 == 0 and (year % 100 != 0 or year % 400 == 0)


check_date = datetime(1996, 12, 15)
if is_leap_year(check_date):
  print(check_date.strftime("%Y"), "is a leap year")
else:
  print(check_date.strftime("%Y"), "is not a leap year")

1996 is a leap year


---
### Exercise 5.3 - display parts of date/time

Write a function that will take a date and print from it the following:  
*  given date and time  
*  year (in full) 
*  month (as a number) 
*  month name 
*  week number of the year  
*  weekday of the week (as a number)  
*  day of year  
*  day of the month  
*  day of week (as a name)  

*Hint:  see w3Schools for [strftime reference](https://www.w3schools.com/python/python_datetime.asp)*



In [ ]:
from datetime import datetime

def display_datetime(dt):
  # change the formatting following the tutorial: https://stackabuse.com/how-to-format-dates-in-python/
  # you can also use the properties of the datetime object (year, month, day)
  print(dt.strftime("%b %d %Y %H:%M:%S"))
  print(dt.strftime("%Y"))
  print(dt.strftime("%m"))
  print(dt.strftime("%B"))
  print(dt.strftime("%W"))
  print(dt.strftime("%w"))
  print(dt.strftime("%j"))
  print(dt.strftime("%d"))
  print(dt.strftime("%A"))


now = datetime.now()
display_datetime(now)

Apr 14 2021 18:53:08
2021
04
April
15
3
104
14
Wednesday


---
### Exercise 5.4 - next 6 days

Write a function that will take a date and generate the dates of the following 6 days.  It should print each of the 6 days in the form:  

DAY OF THE WEEK, MONTH NAME, FULL YEAR

Hint:  to generate the days, use the function datetime.timedelta(), there is a good example of its use [here](https://www.geeksforgeeks.org/python-datetime-timedelta-function/)  

In [ ]:
import datetime
def generate_future_date():
  today = datetime.date.today()
  for i in range(0, 6):
      today2 = today + datetime.timedelta(days=i)
      print(today2.strftime("%A, %B %Y"))
generate_future_date()

Wednesday, April 2021
Thursday, April 2021
Friday, April 2021
Saturday, April 2021
Sunday, April 2021
Monday, April 2021


###  Exercise 6 - Merging & Joining data + Joining data from 2 different sources
                  Create a function that does all the wrangling

Create a function called **get_income_df()** which reads the three data sheets, creates the new versions of `county_level_df` and `state_level_df` and combines the new versions.

This function should:

*  read the three sheets into the dataframes `county_level_df`, `state_level_df` and `income_new_df`
*  clean up the Population column of `county_level_df` so that it has the name Pop
*  clean up the Pop column of `state_level_df` so that it contains values that are whole numbers of 1000s
*  combine `county_level_df` and `state_level_df` with join='inner', ignoring index
*  append income_new_df to the combined dataframe  
*  SORT the final dataframe on Income in descending order
*  return the final dataframe

**Test output**:  
dataframe shape = (33, 4)   
first row = 10	MD	25	73	98     
last row = 7	TX	40	75	5



In [ ]:
import pandas as pd
#read the three sheets into the dataframes county_level_df, state_level_df and income_new_df

df1 = pd.read_excel('https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/Income-Data.xlsx?raw=true')
df2 = pd.read_excel('https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/Income-Data.xlsx?raw=true', sheet_name = 'state-level')
df3 = pd.read_excel('https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/Income-Data.xlsx?raw=true', sheet_name = 'income')

display((df1['Income'].max(), df1['Income'].min()))
display((df2['Income'].max(), df2['Income'].min()))
display((df3['Income'].max(), df3['Income'].min()))



def get_income_df(county_level_df, state_level_df, income_new_df):
  #clean up the Population column of county_level_df so that it has the name Pop
  county_level_df['Pop'] = county_level_df['Population']
  county_level_df.drop('Population', axis = 1)

  state_level_df['Pop'] = state_level_df['Pop']/1000

  combined_df = pd.concat([county_level_df, state_level_df], join='inner', ignore_index=True)
  final_df = combined_df.append(income_new_df, ignore_index=True)
  return final_df.sort_values(by=['Income'], ascending=False)


answer = get_income_df(df1, df2, df3)
display(answer.shape)
display(answer.head(1))
display(answer.tail(1))

answer
#dif answer to test output

(81, 22)

(78, 35)

(78, 35)

(33, 4)

State  Age  Income   Pop
15    MD   22      81  89.0

State  Age  Income   Pop
8    MD   27      22  61.0

State  Age  Income     Pop
15    MD   22      81  89.000
31    CA   29      78  10.000
27    NY   29      78  11.000
22    NY   34      78  18.142
14    MD   31      78  34.000
7     TX   40      75   5.000
11    MD   29      75  64.000
16    MD   25      73  21.000
10    MD   25      73  98.000
25    IN   32      69  23.000
23    TX   32      69  70.000
9     MD   23      69   5.000
19    MD   29      69  10.343
29    MD   34      67  81.000
21    CA   35      67  29.587
32    NY   34      67   7.000
13    MD   25      66  24.000
6     TX   35      66  82.000
0     TX   34      65  72.000
12    MD   24      65  36.000
3     TX   36      65  54.000
17    MD   30      62  21.000
5     TX   25      62  28.000
28    TX   35      54  27.000
18    TX   32      54  23.543
26    CA   35      54  54.000
4     TX   42      53  11.000
2     TX   23      46  25.000
1     TX   42      45  33.000
30    IN   41      35  24.000
20    IN   41      35   5.231
24    MD   41      35  33.000
8     MD   27      22  61.000

### Exercise 7 - Resetting Data Values
                Group by state and calculate statistics

Run the function get_income_df() and store the result in a new dataframe called **income_df**.

Group income_df by State and calculate some statistics on the groups.

Add a description, in the cell below, of what the statistics show and why in the text box.

State     Income        
              mean min max
0    CA  66.333333  54  78
1    IN  46.333333  35  69
2    MD  64.230769  22  81
3    NY  74.333333  67  78
4    TX  59.454545  45  75

In [ ]:
income_df = get_income_df(df1, df2, df3)
grouped1 = income_df.groupby('State').agg({'Age': ['mean', 'min', 'max']}).reset_index()
grouped2 = income_df.groupby('State').agg({'Income': ['mean', 'min', 'max']}).reset_index()
display(grouped2)

#shows the mean, min and max age in each state.
#shows the mean, min and max income in each state.


# income_df

State     Income        
              mean min max
0    CA  66.333333  54  78
1    IN  46.333333  35  69
2    MD  64.230769  22  81
3    NY  74.333333  67  78
4    TX  59.454545  45  75